library/pkparse.c

```diff
@@ -985,8 +983,7 @@ static int pk_parse_key_pkcs8_encrypted_der(
     if( ( ret = mbedtls_asn1_get_tag( &p, end, &len, MBEDTLS_ASN1_OCTET_STRING ) ) != 0 )
         return( MBEDTLS_ERR_PK_KEY_INVALID_FORMAT + ret );
```

Comment:
Good point. The ASN.1 parsing functions guarantee that we don't over-read, but it might happen that the provided keybuffer contains garbage after what we have parsed. While I don't think this is dangerous here, it would be better to check that `end == p + len` instead of setting `end = p + len`. @mpg Do you agree? Is there any reason why we would like to tolerate additional data at the end of a DER encoded key?

Reply:
@mpg @andresag01 There are multiple occasions in `pkparse.c` where this check is omitted. If we agree it's an issue, I'd therefore suggest to open a separate PR for it. 

Besides, we seem to be missing calls to `mbedtls_ecp_keypair_free` at https://github.com/ARMmbed/mbedtls/blob/development/library/pkparse.c#L802, https://github.com/ARMmbed/mbedtls/blob/development/library/pkparse.c#L805 and https://github.com/ARMmbed/mbedtls/blob/development/library/pkparse.c#L816 - would you agree?

Reply:
@hanno-arm: I agree that in the first two cases there is a check missing to ensure that `end == p` at the end of parsing. In the third case, I did not investigate enough, but it also looks like there is a check missing inside `pk_get_ecpubkey()`, but to be honest I am not sure because the code also does `*p = (unsigned char *) end;`

Reply:
@andresag01 I am a bit reluctant to change the behavior of the code here to mandate that the certificate must span the whole buffer. I'll open a ticket on it, as well as another one regarding the missing calls to `ecp_keypair_free`.

Reply:
@andresag01 See IOTSSL-1778 and IOTSSL-1779.